rgd/geodata/tests/test_api.py

```diff
@@ -0,0 +1,136 @@
+
+@pytest.mark.django_db(transaction=True)
+def test_download_file(client, arbitrary_file):
+    model = 'ArbitraryFile'
+    id = arbitrary_file.id
+    field = 'file'
+    result = client.get(f'/api/geodata/download/{model}/{id}/{field}')
+    assert result.status_code == 200
+    with pytest.raises(AttributeError):
+        # Test bad model
+        client.get('/api/geodata/download/Foo/0/file')
+    with pytest.raises(AttributeError):
+        # Test good model, bad field
+        client.get(f'/api/geodata/download/{model}/{id}/foo')
+
+
+@pytest.mark.django_db(transaction=True)
+def test_get_status(client, elevation_image):
+    model = 'ImageFile'
+    id = elevation_image.image_file.imagefile.id
+    result = client.get(f'/api/geodata/status/{model}/{id}')
+    assert result.status_code == 200
+    with pytest.raises(AttributeError):
+        client.get(f'/api/geodata/status/Foo/{id}')
+
+
+@pytest.mark.django_db(transaction=True)
+def test_download_arbitry_file(client, arbitrary_file):
+    pk = arbitrary_file.pk
+    result = client.get(f'/api/geodata/common/arbitrary_file/{pk}/data')
+    assert result.status_code == 302 or result.status_code == 200
+
+
+@pytest.mark.django_db(transaction=True)
+def test_download_image_entry_file(client, elevation_image):
+    pk = elevation_image.pk
+    result = client.get(f'/api/geodata/imagery/image_entry/{pk}/data')
+    assert result.status_code == 302 or result.status_code == 200
+
+
+@pytest.mark.django_db(transaction=True)
+def test_get_arbitrary_file(client, arbitrary_file):
+    pk = arbitrary_file.pk
+    content = json.loads(client.get(f'/api/geodata/common/arbitrary_file/{pk}/').content)
+    assert content
+    # Check that a hyperlink is given to the file data
+    # NOTE: tried using the URLValidator from django but it thinks this URL is invalid
+    assert content['file'].startswith('http')
+
+
+@pytest.mark.django_db(transaction=True)
+def test_get_spatial_entry(client, elevation_raster):
+    """Test individual GET for SpatialEntry model."""
+    pk = elevation_raster.rastermetaentry.spatial_id
+    content = json.loads(client.get(f'/api/geodata/common/spatial_entry/{pk}/').content)
+    assert content
+    assert content['footprint']
+    assert content['outline']
+
+
+@pytest.mark.django_db(transaction=True)
+def test_create_get_subsampled_image(client, elevation_image):
+    """Test POST and GET for SubsampledImage model."""
+    client.post(
+        '/api/geodata/imagery/subsample',
+        {
+            'source_image': elevation_image.id,
+            'sample_type': 'pixel box',
+            'sample_parameters': {'umax': 100, 'umin': 0, 'vmax': 200, 'vmin': 0},
+        },
+    )
+    # TODO: wait for the task kicked of by POST to complete?
+    sub = models.imagery.SubsampledImage.objects.get(source_image=elevation_image.id)
+    assert sub.data
+    # Test the GET
+    content = json.loads(client.get(f'/api/geodata/imagery/subsample/{sub.id}/').content)
+    assert content
```

Comment:
This POST endpoint launches a celery task. How can we set up this test to wait on that task to complete?

Normally, the CELERY_EAGER flags capture the execution but in this case, since we are hitting the POST endpoint through the client it becomes async with the test execution.

Reply:
@mcovalt or @manthey, any ideas on how to resolve this?

I'd like to test the endpoint itself, but I could change this to create a SubsampledImage model directly.

Reply:
Nevermind, I figured this out. I was looking at the wrong issue and it turns out you have to use `json.dumps()` for any internal dicts in the payload.